crates/ty_python_semantic/src/types.rs

```diff
@@ -249,6 +249,37 @@ pub(crate) type NormalizedVisitor<'db> = TypeTransformer<'db, Normalized>;
 #[derive(Debug)]
 pub(crate) struct Normalized;
 
+/// A [`TypeTransformer`] that is used in `recursive_type_normalized` methods.
+/// Calling [`Type::recursive_type_normalized`] will normalize the recursive type.
+/// A recursive type here means a type that contains a `Divergent` type.
+/// Normalizing recursive types allows recursive type inference for divergent functions to converge.
+pub(crate) struct RecursiveTypeNormalizedVisitor<'db> {
+    transformer: TypeTransformer<'db, Normalized>,
+    div: Type<'db>,
+}
+
+impl<'db> RecursiveTypeNormalizedVisitor<'db> {
+    fn new(div: Type<'db>) -> Self {
+        debug_assert!(matches!(div, Type::Dynamic(DynamicType::Divergent(_))));
+        Self {
+            transformer: NormalizedVisitor::default(),
+            div,
+        }
+    }
+
+    fn visit(&self, item: Type<'db>, func: impl FnOnce() -> Type<'db>) -> Type<'db> {
+        self.transformer.visit(item, func)
+    }
+
+    fn visit_no_shift(&self, item: Type<'db>, func: impl FnOnce() -> Type<'db>) -> Type<'db> {
+        self.transformer.visit_no_shift(item, func)
+    }
+
```

Comment:
I think we can track `level` in this visitor (we also don't need level, a single bool if nested shouold be enough)
```suggestion
    fn visit(&self, item: Type<'db>, func: impl FnOnce() -> Type<'db>) -> Type<'db> {
        self.transformer.visit(item, || {
            self.nested = true;
            let res = func();
            self.nested = false;
            res
        })
    }

    fn visit_no_shift(&self, item: Type<'db>, func: impl FnOnce() -> Type<'db>) -> Type<'db> {
        self.transformer.visit_no_shift(item, || {
            self.nested = true;
            let res = func();
            self.nested = false;
            res
        })
    }

```

We could probably even get away by simply always setting `nested` to `true` (outside the callback)